File: Priority_Queue/1642. Furthest Building You Can Reach/1642. Furthest Building You Can Reach.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        int n = heights.length;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int i = 1; i < n; i++) {
            if (heights[i] <= heights[i-1]) continue;
            pq.offer(heights[i] - heights[i-1]);
            if (pq.size() > ladders) {
                if (pq.peek() > bricks) return i-1;
                bricks -= pq.poll();
            }
        }
        return n-1;
    }
}

File: Priority_Queue/1642. Furthest Building You Can Reach/1642. Furthest Building You Can Reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {
        multiset<int>Set;
        
        int count = 0;
        for (int i = 1; i < heights.size(); i++) {
            if (heights[i]-heights[i-1] <= 0) continue;
            
            if (count < ladders) {
                count++;
                Set.insert(heights[i]-heights[i-1]);
            } else {
                Set.insert(heights[i]-heights[i-1]);
                if (bricks < *Set.begin()) return i-1;
                bricks -= *Set.begin();
                Set.erase(Set.begin());
            }
        }
        
        return heights.size()-1;
    }
};
